src/Features/Core/Portable/GenerateMember/GenerateParameterizedMember/AbstractGenerateDeconstructMethodService.State.cs

```diff
@@ -0,0 +1,103 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeGeneration;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.GenerateMember.GenerateParameterizedMember
+{
+    internal partial class AbstractGenerateDeconstructMethodService<TService, TSimpleNameSyntax, TExpressionSyntax, TInvocationExpressionSyntax>
+    {
+        internal new class State :
+            AbstractGenerateParameterizedMemberService<TService, TSimpleNameSyntax, TExpressionSyntax, TInvocationExpressionSyntax>.State
+        {
+            /// <summary>
+            /// Make a State instance representing the Deconstruct method we want to generate.
+            /// The method will be called "Deconstruct". It will be a member of `typeToGenerateIn`.
+            /// Its arguments will be based on `targetVariables`.
+            /// </summary>
+            public static async Task<State> GenerateDeconstructMethodStateAsync(
+                TService service,
+                SemanticDocument document,
+                SyntaxNode targetVariables,
+                INamedTypeSymbol typeToGenerateIn,
+                CancellationToken cancellationToken)
+            {
+                var state = new State();
+                if (await state.TryInitializeMethodAsync(service, document, targetVariables, typeToGenerateIn, cancellationToken).ConfigureAwait(false))
+                {
+                    return state;
+                }
+
+                return null;
```

Comment:
in general, i prefer failure returns coming first, and success mode last.  i.e. `if (!await ...) { return null; } ... return state`.